app/routines/create_user.rb

```diff
@@ -41,19 +29,44 @@ def exec(username:, title: nil, first_name: nil, last_name: nil,
     transfer_errors_from(outputs[:user], {type: :verbatim})
   end
 
-  # Returns a valid, though possibly non-unique, username
-  def get_valid_username(base)
-    base = randomify_username(base) if base.blank?
+  def generate_unique_valid_username(username)
+    return username if username_is_valid?(username) && username_is_unique?(username)
+
+    loop do
+      username = create_random_username
+      break if username_is_unique?(username) || max_attempts_exceeded?
+    end
+
+    username
+  end
+
+  def max_attempts_exceeded?
+    @attempts_count ||= 0
+    @attempts_count += 1
+    did_exceed = @attempts_count > username_max_attempts
+    did_exceed
+  end
```

Comment:
On second thought, I think I'd prefer this logic to be directly in the generate method.  While I'm normally a fan of private methods to make things more clear and readable, since this one is actually mutating state (and doing more than just checking if max attempts exceeded), I'd vote for this to be just inline in the method.